### ISSUE TITLE
would you like to share your pretrained weights of RFDN_L or RFDN for x2?

### ISSUE THREAD

**Issue by ytao-wang** [njulj/RFDN#9]

**Comment by leftdoor** [comment]
and x3. Thank you for your work and sharing.

**Comment by scutlrr** [comment]
and x4. Thank you for your work and sharing.

**Comment by Senwang98** [comment]
Hi, @leftdoor @scutlrr @ytao-wang 
Have you re-trained the RFDN model?
I trained it with 5e-5, and I got 37.90 on Set5 2X.(I don't think use some tricks may gain 0.15 PSNR)
Have you reproducted the results in the paper already?

**Comment by scutlrr** [comment]
@1187697147 I got about 37.99db on Set5 2x.

**Comment by 1187697147** [comment]
Hi, @scutlrr 
Thanks for your reply! Can you give more details about your train setting?
For me, I used github repo `EDSR-PyTorch` to train RFDN. Firstly, I copy CCA from `IMDN` to replace ESA.
Then, I set batch_size = 16 and train the model for 1000 epoches and use lr = 5e-4(--decay 200-400-600-800)
This is my ```block/RFDB``` code:
```
class RFDB(nn.Module):
    def __init__(self, in_channels, distillation_rate=0.25):
        super(RFDB, self).__init__()
        self.dc = self.distilled_channels = in_channels//2
        self.rc = self.remaining_channels = in_channels
        self.c1_d = conv_layer(in_channels, self.dc, 1)
        self.c1_r = conv_layer(in_channels, self.rc, 3)
        self.c2_d = conv_layer(self.remaining_channels, self.dc, 1)
        self.c2_r = conv_layer(self.remaining_channels, self.rc, 3)
        self.c3_d = conv_layer(self.remaining_channels, self.dc, 1)
        self.c3_r = conv_layer(self.remaining_channels, self.rc, 3)
        self.c4 = conv_layer(self.remaining_channels, self.dc, 3)
        self.act = activation('lrelu', neg_slope=0.05)
        self.c5 = conv_layer(self.dc*4, in_channels, 1)
        # self.esa = ESA(in_channels, nn.Conv2d)
        # self.cca = CCALayer(in_channels)

    def forward(self, input):
        distilled_c1 = self.act(self.c1_d(input))
        r_c1 = (self.c1_r(input))
        r_c1 = self.act(r_c1+input)

        distilled_c2 = self.act(self.c2_d(r_c1))
        r_c2 = (self.c2_r(r_c1))
        r_c2 = self.act(r_c2+r_c1)

        distilled_c3 = self.act(self.c3_d(r_c2))
        r_c3 = (self.c3_r(r_c2))
        r_c3 = self.act(r_c3+r_c2)

        r_c4 = self.act(self.c4(r_c3))

        out = torch.cat([distilled_c1, distilled_c2, distilled_c3, r_c4], dim=1)
        # out_fused = self.esa(self.c5(out)) 
        out_fused = self.cca(self.c5(out)) 

        return out_fused
```
Can you share the codes you used to train?

**Comment by ytao-wang** [comment]
> Hi, @leftdoor @scutlrr @ytao-wang
> Have you re-trained the RFDN model?
> I trained it with 5e-5, and I got 37.90 on Set5 2X.(I don't think use some tricks may gain 0.15 PSNR)
> Have you reproducted the results in the paper already?

Hi, I trained the RFDN_L with params 625,962. I set batch_size = 64, patch_size = 128 and lr = 5e-4(--decay 200-400-600-800) as the paper, and I got 38.03 on Set 5 x2.

**Comment by Senwang98** [comment]
Hi, @ytao-wang 
I thanks for your reply too!
My params of RFDN-L is **619662**, this is strange!
I have shared the _RFDB_ and _CCA_ used in my codes, so do you think is there any wrong with my codes?
Or
can you share the codes you used to trained, I think it is acceptable that the training result is 0.05 lower than the paper.

**Comment by ytao-wang** [comment]
> Hi, @ytao-wang
> I thanks for your reply too!
> My params of RFDN-L is **619662**, this is strange!
> I have shared the _RFDB_ and _CCA_ used in my codes, so do you think is there any wrong with my codes?
> Or
> can you share the codes you used to trained, I think it is acceptable that the training result is 0.05 lower than the paper.

Hi, you should set reduction = 4 in CCAlayer ([#5 (comment)](https://github.com/njulj/RFDN/issues/5#issuecomment-710090699))

**Comment by Senwang98** [comment]
Hi, @ytao-wang 
Thank you very much, maybe I should have seen issues next time!

**Comment by ytao-wang** [comment]
@1187697147 
Hi, I got  indicators of 5 datasets (Set5, Set14, B100, Urban100, Manga109) as follows:
38.03/0.9606
33.65/0.9183 
32.17/0.8996
32.15/0.9282
38.88/0.9772
I think there's a big difference.

**Comment by Senwang98** [comment]
@ytao-wang 
Yes, I think this is normal. Urban100 is often hard to reproduct Have you used training tricks?
The training of SR is usually not 1000 epoch, but 2000 epoch instead which is not reported in SR papers

**Comment by scutlrr** [comment]
> @1187697147
> Hi, I got indicators of 5 datasets (Set5, Set14, B100, Urban100, Manga109) as follows:
> 38.03/0.9606
> 33.65/0.9183
> 32.17/0.8996
> 32.15/0.9282
> 38.88/0.9772
> I think there's a big difference.

Hi, have you reimplemented the experiment and published the paper?  ~-~

**Comment by Senwang98** [comment]
@scutlrr 
I didn't get what you say.

**Comment by scutlrr** [comment]
@ytao-wang 
Whether you have made improvements based on the thesis experiment and then published a new paper.

**Comment by scutlrr** [comment]
> @scutlrr
> I didn't get what you say.

sorry, I got it wrong

**Comment by 1187697147** [comment]
Hi, @scutlrr  @ytao-wang 
I have met one strange problem and I have nobody for any suggestions. 
I am writing paper about SISR, my result on BD model is much higher than any previous work. 
I am not sure if I have generated wrong BD training data.
Have you met similiar situation??

**Comment by kk6398** [comment]
hello, 
      I have seen the dialogue above , you are really great and successful, but I am the green hand on SR, and I meet two questions: 
      i) I trained the RFDN_L with params only 417.212k and I use the torchsummary.py to measure it.   That I  use the torchsummary.py to test other model like IMDN is correct, so I don't know what wrongs occur.
     ii)I set batch_size = 64, patch_size = 128 and lr = 5e-4(--decay 200-400-600-800), but I get the psnr on Set 5 x2, only 37.76...  I check the all details, but I really don't find any wrongs.

Could you help me?    /cry/....  Thank you a lot!!!

**Comment by ytao-wang** [comment]
> hello,
> I have seen the dialogue above , you are really great and successful, but I am the green hand on SR, and I meet two questions:
> i) I trained the RFDN_L with params only 417.212k and I use the torchsummary.py to measure it. That I use the torchsummary.py to test other model like IMDN is correct, so I don't know what wrongs occur.
> ii)I set batch_size = 64, patch_size = 128 and lr = 5e-4(--decay 200-400-600-800), but I get the psnr on Set 5 x2, only 37.76... I check the all details, but I really don't find any wrongs.
> 
> Could you help me? /cry/.... Thank you a lot!!!

**Comment by Senwang98** [comment]
@kk6398 
Hi, please use CCA layer instead of ESA.  In CCA, set r=4 instead of 16 which is commonly used in channel attention.
Then you will get right paramters(by the way, the way to calculate the params is only one line code, why use torchsummary?)

**Comment by kk6398** [comment]
> @kk6398
> Hi, please use CCA layer instead of ESA. In CCA, set r=4 instead of 16 which is commonly used in channel attention.
> Then you will get right paramters(by the way, the way to calculate the params is only one line code, why use torchsummary?)

Thank you for your reply sincerely. I have modified as you mentioned. However, to calculate  parameters of the network only need one line code , could you provide the  specific code?  thanku thanku thanku!

**Comment by Senwang98** [comment]
Hi,@kk6398
print('Total params: %.3fM' % (sum(p.numel() for p in _model.parameters())/1000000.0))

**Comment by kk6398** [comment]
> Hi,@kk6398
> print('Total params: %.3fM' % (sum(p.numel() for p in _model.parameters())/1000000.0))

Thanks a lot. I have tried the code that you mentioned and I place it in the main.py of the EDSR:
        `
def main():
    global model
    if args.data_test == ['video']:
        from videotester import VideoTester
        model = model.Model(args, checkpoint)
        t = VideoTester(args, model, checkpoint)
        t.test()
    else:
        if checkpoint.ok:
            loader = data.Data(args)
            _model = model.Model(args, checkpoint)
            _loss = loss.Loss(args, checkpoint) if not args.test_only else None
            t = Trainer(args, loader, _model, _loss, checkpoint)
            print('Total params: %.3fM' % (sum(p.numel() for p in _model.parameters())/1000000.0))
            while not t.terminate():
                t.train()
                t.test()
            checkpoint.done()`
    Then, I run demo.sh(RFDN: The whole code as you share above(CCA, reduction = 4)), but the final params is still only 395660(scaleX2) and the result of PSNR is very low . I check the code several times, there is no difference between us. So I am so hopeless that I don't know which part is wrong. Could you help me ?  Thanku thanku thanku from the bottom of my heart.

**Comment by Senwang98** [comment]
@kk6398 
Sorry, maybe you need to debug by yourself since we have given you all the code. I guess you need to check some super-param such as layer numbers and etc.

**Comment by kk6398** [comment]
> @kk6398
> Sorry, maybe you need to debug by yourself since we have given you all the code. I guess you need to check some super-param such as layer numbers and etc.

     Thank you for your reply, according your said, I find some wrong. Thankyou thankyou thankyou. I have run the identical params of RFDNx2 as the paper. And I am running the results. Xixi, Thank you again.

**Comment by jmhhenlihai** [comment]
> @1187697147
> Hi, I got indicators of 5 datasets (Set5, Set14, B100, Urban100, Manga109) as follows:
> 38.03/0.9606
> 33.65/0.9183
> 32.17/0.8996
> 32.15/0.9282
> 38.88/0.9772
> I think there's a big difference.

would you like to share your pretrained weights of RFDN for x2 and RFDN for x3?

**Comment by liang-zhenhuan** [comment]
> > @1187697147
> > Hi, I got indicators of 5 datasets (Set5, Set14, B100, Urban100, Manga109) as follows:
> > 38.03/0.9606
> > 33.65/0.9183
> > 32.17/0.8996
> > 32.15/0.9282
> > 38.88/0.9772
> > I think there's a big difference.
> 
> would you like to share your pretrained weights of RFDN for x2 and RFDN for x3?

你好，我觉得你复现的结果和原文结果非常接近，想请问您可否分享一下您的训练代码或者加我QQ836036016，我请教您一下参数设置的问题。感谢您的帮助！

**Comment by jmhhenlihai** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by ytao-wang** [comment]
> > > @1187697147
> > > Hi, I got indicators of 5 datasets (Set5, Set14, B100, Urban100, Manga109) as follows:
> > > 38.03/0.9606
> > > 33.65/0.9183
> > > 32.17/0.8996
> > > 32.15/0.9282
> > > 38.88/0.9772
> > > I think there's a big difference.
> > 
> > 
> > would you like to share your pretrained weights of RFDN for x2 and RFDN for x3?
> 
> 你好，我觉得你复现的结果和原文结果非常接近，想请问您可否分享一下您的训练代码或者加我QQ836036016，我请教您一下参数设置的问题。感谢您的帮助！
你好，这是RFDN-L的网络框架代码，注意：nf=52，num_modules=6
`class RFDN(nn.Module):
    def __init__(self, in_nc=3, nf=52, num_modules=6, out_nc=3, upscale=2):
        super(RFDN, self).__init__()
        # self.scale_idx = 0
        self.fea_conv = B.conv_layer(in_nc, nf, kernel_size=3)

        self.B1 = B.RFDB(in_channels=nf)
        self.B2 = B.RFDB(in_channels=nf)
        self.B3 = B.RFDB(in_channels=nf)
        self.B4 = B.RFDB(in_channels=nf)
        self.B5 = B.RFDB(in_channels=nf)
        self.B6 = B.RFDB(in_channels=nf)
        self.c = B.conv_block(nf * num_modules, nf, kernel_size=1, act_type='lrelu')

        self.LR_conv = B.conv_layer(nf, nf, kernel_size=3)

        upsample_block = B.pixelshuffle_block
        self.upsampler = upsample_block(nf, out_nc, upscale_factor=upscale)

    def forward(self, input):
        out_fea = self.fea_conv(input)
        out_B1 = self.B1(out_fea)
        out_B2 = self.B2(out_B1)
        out_B3 = self.B3(out_B2)
        out_B4 = self.B4(out_B3)
        out_B5 = self.B5(out_B4)
        out_B6 = self.B6(out_B5)

        out_B = self.c(torch.cat([out_B1, out_B2, out_B3, out_B4, out_B5, out_B6], dim=1))
        out_lr = self.LR_conv(out_B) + out_fea

        output = self.upsampler(out_lr)

        return output`
通过torchstat包，我计算得到模型参数：625,962(upscale=2)，632,997(upscale=3)，642,846(upscale=4)，上述结果需将RFDB中的CCALayer（计算RFDN的参数与论文的一致）替换成ESA。训练时候的设置我上面回复有提及。

**Comment by ytao-wang** [comment]
> > > @1187697147
> > > Hi, I got indicators of 5 datasets (Set5, Set14, B100, Urban100, Manga109) as follows:
> > > 38.03/0.9606
> > > 33.65/0.9183
> > > 32.17/0.8996
> > > 32.15/0.9282
> > > 38.88/0.9772
> > > I think there's a big difference.
> > 
> > 
> > would you like to share your pretrained weights of RFDN for x2 and RFDN for x3?
> 
> 你好，我觉得你复现的结果和原文结果非常接近，想请问您可否分享一下您的训练代码或者加我QQ836036016，我请教您一下参数设置的问题。感谢您的帮助！

**Comment by Senwang98** [comment]
@ytao-wang 
I give up SR this field and fail to help you, sorry!

**Comment by ytao-wang** [comment]
@liang-zhenhuan 

下面是RFDN-L的网络框架代码：

`class RFDN(nn.Module):
    def __init__(self, in_nc=3, nf=52, num_modules=6, out_nc=3, upscale=3):
        super(RFDN, self).__init__()
        # self.scale_idx = 0
        self.fea_conv = B.conv_layer(in_nc, nf, kernel_size=3)

        self.B1 = B.RFDB(in_channels=nf)
        self.B2 = B.RFDB(in_channels=nf)
        self.B3 = B.RFDB(in_channels=nf)
        self.B4 = B.RFDB(in_channels=nf)
        self.B5 = B.RFDB(in_channels=nf)
        self.B6 = B.RFDB(in_channels=nf)
        self.c = B.conv_block(nf * num_modules, nf, kernel_size=1, act_type='lrelu')

        self.LR_conv = B.conv_layer(nf, nf, kernel_size=3)

        upsample_block = B.pixelshuffle_block
        self.upsampler = upsample_block(nf, out_nc, upscale_factor=upscale)

    def forward(self, input):
        out_fea = self.fea_conv(input)
        out_B1 = self.B1(out_fea)
        out_B2 = self.B2(out_B1)
        out_B3 = self.B3(out_B2)
        out_B4 = self.B4(out_B3)
        out_B5 = self.B5(out_B4)
        out_B6 = self.B6(out_B5)

        out_B = self.c(torch.cat([out_B1, out_B2, out_B3, out_B4, out_B5, out_B6], dim=1))
        out_lr = self.LR_conv(out_B) + out_fea

        output = self.upsampler(out_lr)

        return output`
注意设置nf=52，num_modules=6。通过torchstat包，我计算得到模型参数：625,962(upscale=2)，632,997(upscale=3)，642,846(upscale=4)，需要将RFDB中的CCALayer替换成ESA。训练的设置可以看我之前的回复。

**Comment by Senwang98** [comment]
@liang-zhenhuan
这个我记得只有CA的reduction=4这个坑，其他的非常简单复现
（加qq解答问题可还行。。。。

**Comment by Liiiiaictx** [comment]
@Senwang98 大佬，这个RFDN怎么基于EDSR框架训练的，需要把这个仓库的文件拷贝到EDSR仓库吗？其次RFDN的环境配置是什么样的？

**Comment by jmhhenlihai** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by Liiiiaictx** [comment]
@ytao-wang 大佬，这个RFDN怎么基于EDSR框架训练的，需要把这个仓库的文件拷贝到EDSR仓库吗？其次RFDN的环境配置是什么样的？
